MCU Doc fixes (#1615)
slint-ui/slint: api/rs/slint/mcu.md

# Slint on Microcontrollers (MCU)

This document explains how to use Slint to develop a UI on a MCU in a bare metal environment.

## Install Toolchain / Hardware Abstraction Layer (HAL)

Writing an application in Rust that runs on a MCU requires the following software components:

* You need to install a Rust toolchain to cross-compile to the target architecture.
* You need to locate and select the correct HAL crates and drivers, and depend on them in your `Cargo.toml`.
* You need to install tools for flashing and debugging your code on the device.

Covering these is out of scope for this document, but we recommend reading the [Rust Embedded Book](https://docs.rust-embedded.org/book/),
as well as the curated list of [Awesome Embedded Rust](https://github.com/rust-embedded/awesome-embedded-rust) for a wide range of different
crates, tools and training materials. These resources should guide you through the initial setup and often come with "hello world" examples
to get started with your device.

> In addition, Slint requires a nightly version of Rust. This is due to the fact that the support for using a custom global allocator in a bare metal

Reviewer: ```suggestion
In order to set a global allocator, required by Slint, you will need a **nightly** version of Rust. This is due to the fact that the support for using a custom global allocator in a bare metal
```

(Slint itself don't require nightly)

environment with `#![no_std]` has not been stabilized yet (see [#51540](https://github.com/rust-lang/rust/issues/51540) or
[#66741](https://github.com/rust-lang/rust/issues/66741) for tracking issues).

In the following sections we assume that your setup is complete and you have a non-graphical skeleton Rust program running on your MCU.

## Changes to `Cargo.toml`

Start by adding a dependency to the `slint` crate to your `Cargo.toml`:

```toml
[dependencies]

> ... your other dependencies

Reviewer: ```suggestion
## ... your other dependencies
```

I think two # escapes in one #

Author: Ahh! yes, that works, thanks!

[dependencies.slint]
version = "0.3.0"
default-features = false
features = ["compat-0.3.0", "unsafe-single-threaded", "libm"]

[build-dependencies]

> slint-build = version = "0.3.0"

Reviewer: ```suggestion
slint-build = "0.3"
```

```

The default features of the `slint` create are tailored towards hosted environments and includes the "std" feature. In bare metal environments,
you need to disable the default features and select at least the `compat-0.3.0` feature (see [this blog post](https://slint-ui.com/blog/rust-adding-default-cargo-feature.html)
for a longer explanation).

> In addition we select two additional features:

Reviewer 2: Can we drop the "In addition"?

"We select two additional features:"

Author: I'll change this indeed, the tone is all mixed up. It starts with "Start by adding... to your Cargo.toml" and then continues with "we do XYZ".

 * `unsafe-single-threaded`: Slint internally uses Rust's [`thread_local!`](https://doc.rust-lang.org/std/macro.thread_local.html) macro to store global data.
   This feature is only available in the Rust Standard Library (std), which is not available in bare-metal environments. As a fallback, the `unsafe-single-threaded`
   feature will change Slint to use unsafe static for storage. By setting this feature, you guarantee not to use Slint API from a thread other than the main thread,
   or from interrupt handlers.
 * `libm`: MCUs often don't provide hardware support for floating point arithmetic. The `libm` feature enables the use of software emulation, with the help of
   the [libm](https://crates.io/crates/libm) crate.

> Finally, we add `slint-build` as a build dependency in order to compile the `.slint` design files to Rust code.

Reviewer 2: This sounds like another feature we add, not like an additional crate we need.

Author: Well, it's *another* thing in addition to just disabling the default features :-)

## Application Structure

A graphical application in hosted environments is typically composed of at least three different tasks:

 * Receive user input from operation system APIs.
 * React to the input by performing application specific computations.
 * Render an updated user interface and present it on the screen using device-independent operating system APIs.

The operating system provides what is typically called an event loop to connect and schedule these tasks. Slint implements the
task of receiving user input and forwarding it to the user interface layer, as well as rendering the interface to the screen.

In bare metal environments it becomes your responsibility to substitute and connect functionality that is otherwise provided by the operating system:

 * You need to select crates that allow you to initialize the chips that drive peripherals, such as a touch input or display controller.
   Sometimes it may be necessary for you to develop your own drivers.
 * You need to drive the event loop yourself by querying peripherals for input, forwarding input into computational modules of your
   application and instructing Slint to render the user interface.

> In Slint the two primary APIs you need to use to accomplish these tasks the `Platform` trait as well as the `slint::Window` struct.

Reviewer 2: `slint::platform::Platform`

Or 

just `Window`?

Author: Good point, that's inconsistent. I'll use their qualified names and make them links.

In the following sections we are going to cover how to use them and how they integrate into your event loop.

### The `Platform` Trait

> The `slint::platform::Platform` trait defines the interface between Slint and platform APIs typically provided by operating and windowing systems.

Reviewer: ```suggestion
The [`slint::platform::Platform`] trait defines the interface between Slint and platform APIs typically provided by operating and windowing systems.
```

You need to provide an minimal implementation of this trait and call [`slint::platform::set_platform`] before constructing your Slint application.

A minimal implementation needs to cover two functions:

 * `fn create_window_adapter(&self) -> Rc<dyn WindowAdapter + 'static>;`: Implement this function to return an implementation of the `WindowAdapter`

>    trait that will be associated with the Slint components you create. We provide a convenience struct `slint::platform::software_renderer::MinimalSoftwareWindow`

Reviewer: ```suggestion
   trait that will be associated with the Slint components you create. We provide a convenience struct [`slint::platform::software_renderer::MinimalSoftwareWindow`]
```

   that implements this trait.
 * `fn duration_since_start(&self) -> Duration`: In order for animations in `.slint` design files to change properties correctly, Slint needs to know
   how much time has elapsed since two rendered frames. In a bare metal environment you need to provide a source of time. Often the HAL crate of your
   device provides a system timer API for this, which you can query in your impementation.

There are additional functions in the trait that you can implement, for example handling of debug output, a delegated event loop or an interface
to safely deliver events in multi-threaded environments.

> A typical minimal implementation of the `Platform` trait that uses the `MinimalSoftwareWindow` looks like this:

Reviewer: ```suggestion
A typical minimal implementation of the [`Platform`] trait that uses the [`MinimalSoftwareWindow`] looks like this:
```

const DISPLAY_WIDTH: usize = 320;
let mut line_buffer = [slint::platform::software_renderer::Rgb565Pixel(0); DISPLAY_WIDTH];

let mut display = hal::Display::new(/*...*/);

// ... rest of initialization ...

loop {
    // ...
    window.draw_if_needed(|renderer| {
        renderer.render_by_line(DisplayWrapper{
            display: &mut display,
            line_buffer: &mut line_buffer
        });
    });
    // ...
}

```

Hint: In our experience, using then synchronous `DrawTarget::fill_contiguous` function is slow. If
your device is capable of using DMA, you may be able to achieve better performance by using
two line buffers: One buffer to render into with the CPU, while the other buffer is transferred to
the screen using DMA asynchronously.

Hint: If you see wrong colors on your device, it may be necessary to invert the bits in the u16 pixel
representation before sending them to the screen driver.

## Example Implementations

Slint's own examples use a helper crate called `mcu-board-support` that provides implementations of
the `Platform` trait for some MCUs, along with support for touch input and system timers.

You can find the crate in our Git repository at:

>     <https://github.com/slint-ui/slint/tree/master/examples/mcu-board-support>

Reviewer: ```suggestion
<https://github.com/slint-ui/slint/tree/master/examples/mcu-board-support>
```

Remember that in markdown, a 4 space indentation is like a block of code.

Author: Thanks, that explains the build error (odd that I didn't see it locally).

## Changes to `Cargo.toml`

Start by adding a dependency to the `slint` crate to your `Cargo.toml`:

```toml
[dependencies]
slint = { version = "0.3.0", default-features = false, features = ["compat-0.3.0", "unsafe-single-threaded", "libm"] }
# ... other stuff
```

The default features of the `slint` create are tailored towards hosted environments and includes the "std" feature. In bare metal environments,

> you need to disable the default features and select at least the `compat-0.3.0` feature (see [this blog post](https://slint-ui.com/blog/rust-adding-default-cargo-feature.html)

Reviewer: why is this feature not a bullet point like the other?

Author: Good point! Thanks for spotting, I'll fix that.

for a longer explanation).

In addition we select two additional features:

 * `unsafe-single-threaded`: Slint internally uses Rust's [`thread_local!`](https://doc.rust-lang.org/std/macro.thread_local.html) macro to store global data.
   This feature is only available in the Rust Standard Library (std), which is not available in bare-metal environments. As a fallback, the `unsafe-single-threaded`
   feature will change Slint to use unsafe static for storage. By setting this feature, you guarantee not to use Slint API from a thread other than the main thread,
   or from interrupt handlers.

>  * `libm`: MCUs often don't provide hardware support for floating point arithmetic. The `libm` feature enables the use of software emulation, with the help of

Reviewer: afaik, libm is required because these math function are normally provided by the std lib. that's orthogonal to the hardware support

Author: Ok, let me try a different formulation then. You're right, the libm crate also emphasises an eventual inclusion in std.

}
```

> Hint: If you observe that your `build.rs` script is not used, double check that your `Cargo.toml` is using the

Author: Ah, maybe not in this case. I've had it that it didn't run, but I guess I had it placed in the wrong directory. Ok, I'll leave it out :) (just tried on the mcu example).
